primitives/src/merkle_tree/mod.rs

```diff
@@ -341,3 +341,41 @@ pub trait ForgetableMerkleTreeScheme: MerkleTreeScheme {
     /// Return a tree which is entirely forgotten.
     fn from_commitment(commitment: impl Borrow<Self::Commitment>) -> Self;
 }
+
+/// Universal Merkle tree that allows forget/remember elements from the memory
+pub trait ForgetableUniversalMerkleTreeScheme:
+    ForgetableMerkleTreeScheme + UniversalMerkleTreeScheme
+{
+    /// Trim the leaf at position `pos` from memory.
+    ///
+    /// This is similar to [forget](ForgetableMerkleTreeScheme::forget), but it
+    /// may prune even an empty sub-tree at `pos` and will return a
+    /// non-membership proof for the pruned position if it does so. Note
+    /// that an implementation may choose _not_ to prune an empty sub-tree, as
+    /// it may be more efficient to represent an empty sub-tree than a
+    /// forgotten one. In this case,
+    /// [universal_lookup](UniversalMerkleTreeScheme::universal_lookup) may
+    /// return _either_ [LookupResult::NotInMemory] or
+    /// [LookupResult::NotFound] after a successful call to
+    /// [universal_forget](Self::universal_forget). In any case, if this
+    /// function is called for a `pos` which is in memory but not in the
+    /// tree, it will return a non-membership proof.
+    ///
+    /// The return value is the same as if
+    /// [universal_lookup](UniversalMerkleTreeScheme::universal_lookup) were
+    /// called before this call.
+    fn universal_forget(
+        &mut self,
+        pos: Self::Index,
+    ) -> LookupResult<Self::Element, Self::MembershipProof, Self::NonMembershipProof>;
```

Comment:
thanks for the detailed code doc, question: is this more like a convenient API that combines the forget-then-universal-lookup two steps into one? or is this functionally different?